### PR TITLE
Optimize project file decoding

### PR DESCRIPTION
**Pull request by CognitiveDisson**
Hi.
I was going to use your framework to implement a remote cache. But I ran into a blocking problem - it takes my project file about 12 seconds to read from the disk and about 30 seconds to write back. The project is really big - about 13 megabytes, but even for this size such performance is extremely bad. 

I tried to find a issue about this, but did not find it. After that, I looked at the source code and found a lot of performance issues. Some of them I tried to fix with minimal changes. As a result, the project opening time was reduced from 12 to 3 seconds. There are still a few problems left in the code, but the main problems have been fixed. In general, I would recommend to refuse Codable and use dictionaries and Marshal - [benchmarks](https://github.com/bwhiteley/JSONShootout).

## Problem research

I used Xcode Instruments and my [utilitarian class](https://gist.github.com/CognitiveDisson/d04ce89c5ad38ddd46deb31efca2cece) to look for problems.

I made a demo project to check how long it takes to serialize a project file using a simple PropertyListSerialization.
```
    func readPlist(url: URL) {
        do {
            let plistXML = try Data(contentsOf: url)
            var plistData: [String: AnyObject] = [:]
            var propertyListForamt =  PropertyListSerialization.PropertyListFormat.xml
            do {
                plistData = try PropertyListSerialization.propertyList(from: plistXML, options: .mutableContainersAndLeaves, format: &propertyListForamt) as! [String : AnyObject]
                let objects = plistData["objects"] as! [String: AnyObject]
                let classes = plistData["classes"] as! [String: AnyObject]
                let firstObject = objects.first
                let firstClass = classes.first
            } catch {
                print("Error reading plist: \(error), format: \(propertyListForamt)")
            }
        } catch {
            print("error no upload")
        }
    }
```
It turned out that it takes less than a second (very quickly). I understood that the problem is in decoding and serialization.

### Concurrent decoding

First of all, I added concurrent decoding for objects in PBXObjects and implemented synchronization mechanisms for it.

Before: 
```
try objectsDictionaries.forEach { reference, dictionary in
    let object = try PBXObject.parse(reference: reference, dictionary: dictionary, userInfo: decoder.userInfo)
    objects.add(object: object)
}
```

After:
```
let dictionary = NSDictionary(dictionary: objectsDictionaries)
dictionary.enumerateKeysAndObjects(options: .concurrent) { (key, obj, _) in
    guard let reference = key as? String,
        let dictionary = obj as? Dictionary<String, Any>
        else { return }
    let userInfo = decoder.userInfo
    let parser = PBXObjectParser(
        reference: reference,
        userInfo: userInfo
    )
    guard let object = try? parser.parse(dictionary: dictionary)
        else { return }
    objects.add(object: object)
}
```

I chose the standard method `enumerateKeysAndObjects`, but I also tried to [manually select the number of threads](https://gist.github.com/CognitiveDisson/bb88b26262b49b3c78a2ad737024a373). There was no significant difference. This gave about 4 seconds. (  8 seconds instead of 12 ). Instruments showed excellent work parallelization.

### Blind decoding order

I also added very simple optimizations based on the order of parsing - since there are usually more `PBXBuildFile` entries in the project than `PBXTarget` ones, so we should start our blind decoding from them, thus reducing the number of checks. Similar with switches by type. This gave another 1-2 seconds. (6 seconds instead of 12)

### Eliminating unnecessary serialization

After that I found a time consuming decoding here:
```
let objectsDictionary: [String: Any] = try container.decodeIfPresent([String: Any].self, forKey: .objects)
```

I realized that in this case you can convert this project file to serialize into a dictionary and just take a value from it. To make this more accurate, I added a new property to the `ProjectDecodingContext` and read from it directly in the `KeyedDecodingContainer` extension. Currenly this optimization takes effect only during `PBXProj` decoding. For other objects there will be no key provided, but this extra check does not affect their decoding times. ( In the future, it is possible to improve this for nested objects. )

## Result

Finally, I got the cherished 3 seconds. Ideally, I would like about a 1-2 seconds, but this requires global changes (like replacing `Codable` with `Marshal`).

All tests pass (checked many times), and the project parsing is also correct (checked it in my demo project).

I also plan to speed up the project saving performance in the next PR, and I already know what to do.

### CODE REVIEW

**Comment by keith** [comment]
I should have a chance to test this in the next few days, our pbxproj file is 32 mbs so it would be great to see an improvement. I'll have to locally work this into an XcodeGen build though

File: Sources/xcodeproj/Objects/Project/PBXProj.swift [hunk 1]
```diff
@@ -70,10 +70,22 @@ public final class PBXProj: Decodable {
...
         let objectsDictionary: [String: Any] = try container.decodeIfPresent([String: Any].self, forKey: .objects) ?? [:]
         let objectsDictionaries: [String: [String: Any]] = (objectsDictionary as? [String: [String: Any]]) ?? [:]
-        try objectsDictionaries.forEach { reference, dictionary in
-            let object = try PBXObject.parse(reference: reference, dictionary: dictionary, userInfo: decoder.userInfo)
+        
+        let dictionary = NSDictionary(dictionary: objectsDictionaries)
+        dictionary.enumerateKeysAndObjects(options: .concurrent) { (key, obj, _) in
+            guard let reference = key as? String,
+                let dictionary = obj as? Dictionary<String, Any>
+                else { return }
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Project/PBXProj.swift]: If the key is not a string, nor the value is a dictionary, we are skipping one entry here, which will result in a in-memory representation of the project that doesn't match the content in disk. What do you think about assuming that the types are the expected ones?

```swift
let reference = key as! String
let dictionary = obj as! Dictionary<String, Any>
```

Alternatively, we can throw an error that gives us more details about why it failed. The idea is being able to catch those as issues, instead of swallowing them.

File: Sources/xcodeproj/Extensions/Decodable+Dictionary.swift [hunk 2]
```diff
@@ -8,8 +8,10 @@ extension Decodable {
     /// - Parameter jsonDictionary: json dictionary.
     /// - Throws: throws an error if the initialization fails.
     init(jsonDictionary: [String: Any]) throws {
```
Comment (pepibumur) [inline Sources/xcodeproj/Extensions/Decodable+Dictionary.swift]: What do you think about turning this into a static method that is more explicit about the type of decodable that is returning?

```swift
static func projectDecodable(dictionary: [String: Any]) throws -> Decodable
```

File: Sources/xcodeproj/Project/XcodeProj.swift [hunk 3]
```diff
@@ -27,8 +27,8 @@ public final class XcodeProj: Equatable {
             if pbxprojPaths.count == 0 {
                 throw XCodeProjError.pbxprojNotFound(path: path)
             }
-            let pbxProjData = try Data(contentsOf: pbxprojPaths.first!.url)
-            let context = ProjectDecodingContext()
+            let (pbxProjData, jsonDictionary) = try XcodeProj.readPBXProj(url: pbxprojPaths.first!.url)
+            let context = ProjectDecodingContext(jsonDictionary: jsonDictionary)
```
Comment (pepibumur) [inline Sources/xcodeproj/Project/XcodeProj.swift]: I think this one can be optimised as well by passing a reference instead of a value type. If I'm not wrong, Swift creates a copy of the whole `jsonDictionary` when initialising the `ProjectDecodingContext` object.

File: Sources/xcodeproj/Project/XcodeProj.swift [hunk 4]
```diff
@@ -69,6 +69,22 @@ public final class XcodeProj: Equatable {
             lhs.pbxproj == rhs.pbxproj &&
             lhs.sharedData == rhs.sharedData
     }
+    
+    // MARK: - Private
+    
+    private static func readPBXProj(url: URL) throws -> (Data, [String: Any]) {
+        let plistXML = try Data(contentsOf: url)
+        var propertyListForamt =  PropertyListSerialization.PropertyListFormat.xml
```
Comment (pepibumur) [inline Sources/xcodeproj/Project/XcodeProj.swift]: Typo: `propertyListForamt`

File: Sources/xcodeproj/Project/XcodeProj.swift [hunk 5]
```diff
@@ -69,6 +69,22 @@ public final class XcodeProj: Equatable {
             lhs.pbxproj == rhs.pbxproj &&
             lhs.sharedData == rhs.sharedData
     }
+    
+    // MARK: - Private
+    
+    private static func readPBXProj(url: URL) throws -> (Data, [String: Any]) {
```
Comment (pepibumur) [inline Sources/xcodeproj/Project/XcodeProj.swift]: I'd rather pass a `Path` argument to be consistent with the rest of the code.

File: Sources/xcodeproj/Project/XcodeProj.swift [hunk 6]
```diff
@@ -27,8 +27,8 @@ public final class XcodeProj: Equatable {
             if pbxprojPaths.count == 0 {
                 throw XCodeProjError.pbxprojNotFound(path: path)
             }
-            let pbxProjData = try Data(contentsOf: pbxprojPaths.first!.url)
-            let context = ProjectDecodingContext()
+            let (pbxProjData, jsonDictionary) = try XcodeProj.readPBXProj(url: pbxprojPaths.first!.url)
```
Comment (pepibumur) [inline Sources/xcodeproj/Project/XcodeProj.swift]: I'd rename `jsonDictionary` to `pbxProjDictionary`

File: Sources/xcodeproj/Project/XcodeProj.swift (hunk 3 above)
Comment (pepibumur) [reply Sources/xcodeproj/Project/XcodeProj.swift]: Nitpick: I'd also rename the name of the argument to `pbxProjDictionary`

File: Sources/xcodeproj/Utils/Decoders.swift [hunk 7]
```diff
@@ -12,6 +13,10 @@ class PBXObjectReferenceRepository {
     ///   - objects: objects.
     /// - Returns: object reference.
     func getOrCreate(reference: String, objects: PBXObjects) -> PBXObjectReference {
+        lock.lock()
```
Comment (pepibumur) [inline Sources/xcodeproj/Utils/Decoders.swift]: What about using the extension for `NSRecursiveLock` that you introduced?

File: Sources/xcodeproj/Project/XcodeProj.swift [hunk 8]
```diff
@@ -69,6 +69,22 @@ public final class XcodeProj: Equatable {
...
+    private static func readPBXProj(url: URL) throws -> (Data, [String: Any]) {
+        let plistXML = try Data(contentsOf: url)
+        var propertyListForamt =  PropertyListSerialization.PropertyListFormat.xml
+        let serialized = try PropertyListSerialization.propertyList(
+            from: plistXML,
+            options: .mutableContainersAndLeaves,
+            format: &propertyListForamt
+        )
+        guard let dictionary = serialized as? [String: Any] else {
+            return (plistXML, [:])
```
Comment (pepibumur) [inline Sources/xcodeproj/Project/XcodeProj.swift]: Am I wrong in thinking that if the serialised object is not a `[String: Any]` the performance improvements that you introduced will not be executed? I think it's safe to use `as!` here. I doubt that Xcode changes its format to have a key other than a string.

File: Sources/xcodeproj/Extensions/NSRecursiveLock+Sync.swift [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+extension NSRecursiveLock {
+    func sync<T>(closure: () -> (T)) -> T {
```
Comment (pepibumur) [inline Sources/xcodeproj/Extensions/NSRecursiveLock+Sync.swift]: Is this being used?

Comment (beefon) [reply Sources/xcodeproj/Extensions/NSRecursiveLock+Sync.swift]: I saw `whileLocked()` naming across SPM in similar extensions. Perhaps, you might consider using it as well.

File: Sources/xcodeproj/Extensions/Decodable+Dictionary.swift (hunk 2 above)
Comment (pepibumur) [inline Sources/xcodeproj/Extensions/Decodable+Dictionary.swift]: We can get rid of this extension. I searched for it in the project and is only used in a couple of tests, which can be refactored to not use it.

File: Sources/xcodeproj/Utils/Decoders.swift [hunk 10]
```diff
@@ -28,10 +33,12 @@ class ProjectDecodingContext {
 
     /// Objects.
     let objects: PBXObjects
+    let jsonDictionary: [String: Any]
 
-    init() {
+    init(jsonDictionary: [String: Any] = [:]) {
```
Comment (beefon) [inline Sources/xcodeproj/Utils/Decoders.swift]: Do you need to have a default value here?

File: Sources/xcodeproj/Objects/Project/PBXObjectParser.swift [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    let userInfo: [CodingUserInfoKey: Any]
+    
+    init(reference: String, userInfo: [CodingUserInfoKey: Any]) {
+        self.reference = reference
+        self.userInfo = userInfo
+    }
+    
+    // swiftlint:disable function_body_length
+    public func parse(dictionary: [String: Any]) throws -> PBXObject {
+        let decoder = XcodeprojJSONDecoder()
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Project/PBXObjectParser.swift]: Looking at the code I think there's a bug here *(not introduced by the PR)*. The decoder, has a context object with a property `objectReferenceRepository` which is used by the objects when they are being initialised to create the references with other objects. When we create a decoder for each object being parsed, we are passing an empty reference repository.

We can fix it by moving the decoder to an instance variable of `PBXObjectParser`, and change the logic that iterates through the keys and values of the pbxproj to use the same parser instead of creating an instance in every step of the iteration.

File: Sources/xcodeproj/Objects/Project/PBXObjectParser.swift [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        decoder.userInfo = userInfo
+        var mutableDictionary = dictionary
+        mutableDictionary["reference"] = reference
+        let data = try JSONSerialization.data(withJSONObject: mutableDictionary, options: [])
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Project/PBXObjectParser.swift]: This is an unnecessary step that was introduced to leverage the decoders but that we can get rid of in a follow-up PR and get another performance improvement.

File: Sources/xcodeproj/Objects/Project/PBXProj.swift [hunk 13]
```diff
@@ -70,10 +70,22 @@ public final class PBXProj: Decodable {
...
+            let userInfo = decoder.userInfo
+            let parser = PBXObjectParser(
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Project/PBXProj.swift]: I'd move this outside `dictionary.enumerateKeysAndObjects` and change its constructor to be:

```swift
let parser = PBXObjectParser(userInfo: userInfo)
```

Then you can call it for each object:

```swift
parser.parse(reference: reference, dictionary: dictionary)
```

File: Sources/xcodeproj/Extensions/NSRecursiveLock+Sync.swift (hunk 9 above)
Comment (pepibumur) [reply Sources/xcodeproj/Extensions/NSRecursiveLock+Sync.swift]: That's a good idea.

**Comment by pepibumur** [comment]
> I should have a chance to test this in the next few days, our pbxproj file is 34 mbs so it would be great to see an improvement. I'll have to locally work this into an XcodeGen build though

It'd be great @Keith if you could give us some numbers from that project.

File: Sources/xcodeproj/Extensions/Decodable+Dictionary.swift (hunk 2 above)
Comment (CognitiveDisson) [reply Sources/xcodeproj/Extensions/Decodable+Dictionary.swift]: I will move this file to the test folder - `Tests/xcodeprojTests / Extensions`

**Comment by keith** [comment]
Looks like this saves ~4 seconds or ~10% on the largest version of our project 🎉 (I didn't test a ton)

File: Sources/xcodeproj/Project/XcodeProj.swift (hunk 3 above)
Comment (CognitiveDisson) [reply Sources/xcodeproj/Project/XcodeProj.swift]: Fixed

File: Sources/xcodeproj/Utils/Decoders.swift (hunk 10 above)
Comment (CognitiveDisson) [reply Sources/xcodeproj/Utils/Decoders.swift]: Yes. This object is also widely used as a default parameter.

File: Sources/xcodeproj/Objects/Project/PBXObjectParser.swift (hunk 12 above)
Comment (CognitiveDisson) [reply Sources/xcodeproj/Objects/Project/PBXObjectParser.swift]: In follow-up PR

**Comment by CognitiveDisson** [comment]
@keith Strange, I expected a better result. Can you provide more information about the project file? What objects are there most?
If possible, I would like to debug on your project .

@pepibumur I corrected most of the comments. Can I merge?

I implemented using a custom decoder to work with a dictionary. This should take away the extra work. According to the results received a slight but still improved. (from 3 seconds to 2)
I'll try to check on other project files. And make a separate PR.

**Comment by keith** [comment]
I can check in more detail but we have ~700 targets and like 15k files

File: Sources/xcodeproj/Errors/Errors.swift [hunk 14]
```diff
@@ -21,6 +22,8 @@ public enum XCodeProjError: Error, CustomStringConvertible {
             return "The project doesn't contain a .pbxproj file at path: \(path.string)"
         case let .xcworkspaceNotFound(path):
             return "The project doesn't contain a .xcworkspace at path: \(path.string)"
+        case let .pbxprojIncorrectSerializationType(path, expectedType, serialized):
+            return "Can't serialize pbxproj plist at path: \(path.string) to: \(expectedType) but serialized result type is \(type(of: serialized)): \(serialized) "
```
Comment (pepibumur) [inline Sources/xcodeproj/Errors/Errors.swift]: Nitpick: I'd rephrase this one to:

```
Can't serialise \(serialized) into a \(expectedType) for project at path \(path)
```

File: Sources/xcodeproj/Objects/Project/PBXProj.swift [hunk 15]
```diff
@@ -70,10 +70,21 @@ public final class PBXProj: Decodable {
...
-            let object = try PBXObject.parse(reference: reference, dictionary: dictionary, userInfo: decoder.userInfo)
+        
+        let parser = PBXObjectParser(
+            userInfo: decoder.userInfo
+        )
+        let dictionary = NSDictionary(dictionary: objectsDictionaries)
+        dictionary.enumerateKeysAndObjects(options: .concurrent) { (key, obj, _) in
+            let reference = key as! String
+            let dictionary = obj as! Dictionary<String, Any>
+            guard let object = try? parser.parse(
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Project/PBXProj.swift]: I'd not use `try?` here and bubble up the error instead. By doing that we prevent the error from going unnoticed and help us detect if there's something to be supported/fixed.

File: Sources/xcodeproj/Project/XcodeProj.swift [hunk 16]
```diff
@@ -69,6 +75,26 @@ public final class XcodeProj: Equatable {
...
+    
+    private static func readPBXProj(path: Path) throws -> (Data, [String: Any]) {
+        let plistXML = try Data(contentsOf: path.url)
+        var propertyListFormat =  PropertyListSerialization.PropertyListFormat.xml
+        let serialized = try PropertyListSerialization.propertyList(
+            from: plistXML,
+            options: .mutableContainersAndLeaves,
+            format: &propertyListFormat
+        )
+        guard let dictionary = serialized as? [String: Any] else {
```
Comment (pepibumur) [inline Sources/xcodeproj/Project/XcodeProj.swift]: You can safely cast the `serialised` object into a [String: Any] with `as!`. I don't see Xcode replacing the root format of the `pbxproj` file.

File: Sources/xcodeproj/Errors/Errors.swift (hunk 14 above)
Comment (CognitiveDisson) [reply Sources/xcodeproj/Errors/Errors.swift]: Change throw on force cast and remove this error

File: Sources/xcodeproj/Project/XcodeProj.swift (hunk 16 above)
Comment (CognitiveDisson) [reply Sources/xcodeproj/Project/XcodeProj.swift]: I had to add an extension. Method `enumerateKeysAndObjects` doesn't throw error.

**Comment by pepibumur** [comment]
Awesome job @CognitiveDisson 🎉 . I'm merging it.
I've opened [an issue](https://github.com/tuist/xcodeproj/issues/369) to keep track of all the potential performance improvements that could be added to the project.
